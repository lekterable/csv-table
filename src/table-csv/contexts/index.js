import { createContext } from 'react'

export const ModeContext = createContext({ isEditable: false })
export const ContentContext = createContext([
  { content: [], uniqueKeys: [] },
  () => {}
])
