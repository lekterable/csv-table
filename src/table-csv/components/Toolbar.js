import React, { useContext, useState } from 'react'
import { addColumn, addRow, removeColumn, removeRow } from '../actions'
import { ContentContext } from '../contexts'

export default () => {
  const [{ content }, dispatch] = useContext(ContentContext)
  const [header] = content
  const [row, setRow] = useState(1)
  const [column, setColumn] = useState(1)

  return (
    <div data-testid="toolbar" className="toolbar">
      <div className="toolbar__row">
        <input
          data-testid="row-add"
          className="button button--primary"
          type="button"
          value="Add Row"
          onClick={() => dispatch(addRow(Number(row)))}
        />
        <input
          data-testid="row-remove"
          className="button button--primary"
          type="button"
          value="Remove Row"
          onClick={() => dispatch(removeRow(Number(row)))}
        />
        <input
          data-testid="row-input"
          className="input-number"
          type="number"
          value={row}
          min="1"
          max={content.length}
          onChange={e => setRow(e.target.value)}
        />
      </div>
      <div className="toolbar__row">
        <input
          data-testid="column-add"
          className="button button--primary"
          type="button"
          value="Add Column"
          onClick={() => dispatch(addColumn(Number(column)))}
        />
        <input
          data-testid="column-remove"
          className="button button--primary"
          type="button"
          value="Remove Column"
          onClick={() => dispatch(removeColumn(Number(column)))}
        />
        <input
          data-testid="column-input"
          className="input-number"
          type="number"
          value={column}
          min="1"
          max={header ? header.length + 1 : 1}
          onChange={e => setColumn(e.target.value)}
        />
      </div>
    </div>
  )
}
