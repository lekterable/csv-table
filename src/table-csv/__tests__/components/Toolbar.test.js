import React from 'react'
import renderer from 'react-test-renderer'
import { cleanup, fireEvent, render } from 'react-testing-library'
import Toolbar from '../../components/Toolbar'
import { ContentContext } from '../../contexts'

describe('<Toolbar />', () => {
  const withContext = (component, content) => (
    <ContentContext.Provider value={content}>
      {component}
    </ContentContext.Provider>
  )

  afterEach(cleanup)

  it('should match snapshot', () => {
    const component = renderer.create(<Toolbar />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('should add row', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId } = render(
      withContext(<Toolbar />, [{ content: [], uniqueKeys: [] }, dispatchSpy])
    )
    const addRow = getByTestId('row-add')
    const rowInput = getByTestId('row-input')
    fireEvent.change(rowInput, { target: { value: 5 } })
    fireEvent.click(addRow)

    expect(dispatchSpy).toBeCalled()
    expect(dispatchSpy).toBeCalledWith({ rowIndex: 5, type: 'ADD_ROW' })
  })

  it('should remove row', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId } = render(
      withContext(<Toolbar />, [{ content: [], uniqueKeys: [] }, dispatchSpy])
    )
    const removeRow = getByTestId('row-remove')
    const rowInput = getByTestId('row-input')
    fireEvent.change(rowInput, { target: { value: 3 } })
    fireEvent.click(removeRow)

    expect(dispatchSpy).toBeCalled()
    expect(dispatchSpy).toBeCalledWith({ rowIndex: 3, type: 'REMOVE_ROW' })
  })

  it('should add column', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId } = render(
      withContext(<Toolbar />, [{ content: [], uniqueKeys: [] }, dispatchSpy])
    )
    const addColumn = getByTestId('column-add')
    const columnInput = getByTestId('column-input')
    fireEvent.change(columnInput, { target: { value: 4 } })
    fireEvent.click(addColumn)

    expect(dispatchSpy).toBeCalled()
    expect(dispatchSpy).toBeCalledWith({ colIndex: 4, type: 'ADD_COLUMN' })
  })

  it('should remove column', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId } = render(
      withContext(<Toolbar />, [{ content: [], uniqueKeys: [] }, dispatchSpy])
    )
    const removeColumn = getByTestId('column-remove')
    const columnInput = getByTestId('column-input')
    fireEvent.change(columnInput, { target: { value: 2 } })
    fireEvent.click(removeColumn)

    expect(dispatchSpy).toBeCalled()
    expect(dispatchSpy).toBeCalledWith({ colIndex: 2, type: 'REMOVE_COLUMN' })
  })

  it('should set max values', () => {
    const dispatchSpy = jest.fn()
    const content = [
      ['Header 1', 'Header 2', 'Header 3'],
      ['Value 1', 'Value 2', 'Value 3'],
      ['Value 1', 'Value 2', 'Value 3'],
      ['Value 1', 'Value 2', 'Value 3']
    ]

    const { getByTestId } = render(
      withContext(<Toolbar />, [{ content, uniqueKeys: [] }, dispatchSpy])
    )

    const rowInput = getByTestId('row-input')
    const columnInput = getByTestId('column-input')

    expect(Number(rowInput.max)).toEqual(content.length)
    expect(Number(columnInput.max)).toEqual(content[0].length + 1)
  })
})
