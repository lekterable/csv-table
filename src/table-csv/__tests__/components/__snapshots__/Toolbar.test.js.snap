// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar /> should match snapshot 1`] = `
<div
  className="toolbar"
  data-testid="toolbar"
>
  <div
    className="toolbar__row"
  >
    <input
      className="button button--primary"
      data-testid="row-add"
      onClick={[Function]}
      type="button"
      value="Add Row"
    />
    <input
      className="button button--primary"
      data-testid="row-remove"
      onClick={[Function]}
      type="button"
      value="Remove Row"
    />
    <input
      className="input-number"
      data-testid="row-input"
      max={0}
      min="1"
      onChange={[Function]}
      type="number"
      value={1}
    />
  </div>
  <div
    className="toolbar__row"
  >
    <input
      className="button button--primary"
      data-testid="column-add"
      onClick={[Function]}
      type="button"
      value="Add Column"
    />
    <input
      className="button button--primary"
      data-testid="column-remove"
      onClick={[Function]}
      type="button"
      value="Remove Column"
    />
    <input
      className="input-number"
      data-testid="column-input"
      max={1}
      min="1"
      onChange={[Function]}
      type="number"
      value={1}
    />
  </div>
</div>
`;
